sklearn/tree/_splitter.pyx

```diff
@@ -42,6 +43,155 @@ cdef float32_t FEATURE_THRESHOLD = 1e-7
+
+cdef class MinWeightLeafCondition(SplitCondition):
+    def __cinit__(self):
+        self.c.f = min_weight_leaf_condition
+        self.c.e = NULL # min_weight_leaf is stored in splitter, which is already passed to f
+
+cdef bint monotonic_constraint_condition(
+    Splitter splitter,
+    SplitRecord* current_split,
+    intp_t n_missing,
+    bint missing_go_to_left,
+    float64_t lower_bound,
+    float64_t upper_bound,
+    SplitConditionEnv split_condition_env
+) noexcept nogil:
+    if (
+        splitter.with_monotonic_cst and
+        splitter.monotonic_cst[current_split.feature] != 0 and
+        not splitter.criterion.check_monotonicity(
+            splitter.monotonic_cst[current_split.feature],
+            lower_bound,
+            upper_bound,
+        )
+    ):
+        return False
+    
+    return True
+
+cdef class MonotonicConstraintCondition(SplitCondition):
+    def __cinit__(self):
+        self.c.f = monotonic_constraint_condition
+        self.c.e = NULL
+
+# cdef struct HasDataEnv:
+#     int min_samples
+
+# cdef bint has_data_condition(
+#     Splitter splitter,
+#     SplitRecord* current_split,
+#     intp_t n_missing,
+#     bint missing_go_to_left,
+#     float64_t lower_bound,
+#     float64_t upper_bound,
+#     SplitConditionEnv split_condition_env
+# ) noexcept nogil:
+#     cdef HasDataEnv* e = <HasDataEnv*>split_condition_env
+#     return splitter.n_samples >= e.min_samples
+
+# cdef class HasDataCondition(SplitCondition):
+#     def __cinit__(self, int min_samples):
+#         self.c.f = has_data_condition
+#         self.c.e = malloc(sizeof(HasDataEnv))
+#         (<HasDataEnv*>self.c.e).min_samples = min_samples
+    
+#     def __dealloc__(self):
+#         if self.c.e is not NULL:
+#             free(self.c.e)
+        
+#         super.__dealloc__(self)
+
+# cdef struct AlphaRegularityEnv:
+#     float64_t alpha
+
+# cdef bint alpha_regularity_condition(
+#     Splitter splitter,
+#     SplitRecord* current_split,
+#     intp_t n_missing,
+#     bint missing_go_to_left,
+#     float64_t lower_bound,
+#     float64_t upper_bound,
+#     SplitConditionEnv split_condition_env
+# ) noexcept nogil:
+#     cdef AlphaRegularityEnv* e = <AlphaRegularityEnv*>split_condition_env
+
+#     return True
+
+# cdef class AlphaRegularityCondition(SplitCondition):
+#     def __cinit__(self, float64_t alpha):
+#         self.c.f = alpha_regularity_condition
+#         self.c.e = malloc(sizeof(AlphaRegularityEnv))
+#         (<AlphaRegularityEnv*>self.c.e).alpha = alpha
+    
+#     def __dealloc__(self):
+#         if self.c.e is not NULL:
+#             free(self.c.e)
+        
+#         super.__dealloc__(self)
+
+
+# from ._tree cimport Tree
+# cdef class FooTree(Tree):
+#     cdef Splitter splitter
+
+#     def __init__(self):
+#         self.splitter = Splitter(
+#             presplit_conditions = [HasDataCondition(10)],
+#             postsplit_conditions = [AlphaRegularityCondition(0.1)],
+#         )
```

Comment:
Yes. I mainly left them in as a demonstration of the need for the SplitConditionEnv; none of the currently existing SplitConditions require an env because their env is built into the legacy pattern of the omniscient Splitter. So for example if we wanted to do alpha regularity, alpha would be a hyperparameter that would ideally go into a closure. This is a cython implementation of a closure pattern, specifically one that avoids extension types due to the field lookup overhead.